### ISSUE TITLE
How much RAM does data processing consume at peak?

### ISSUE THREAD

**Issue by knuser** [facebookresearch/BLINK#2]
I'm unable to process data on machine with 64GB RAM, my estimation is that `./blink/candidate_retrieval/scripts/get_processed_data.sh data` requires at least 160-190 GB RAM. My estimations for `blink/candidate_retrieval/enrich_data.py` are that it will require even more RAM.

How much RAM should be available to fully process data, load it to SOLR and run demo (section "Use BLINK" from readme)?

**Comment by knuser** [comment]
TL;DR for `./blink/candidate_retrieval/scripts/get_processed_data.sh data` machine should have about 320GB RAM and processing will take up to 3 days.

I have used Google Cloud Platform. At first attempt I choose machine with 256 GB of RAM, but script crashed during processing `blink/candidate_retrieval/enrich_data.py` because insufficient memory available. I have upgraded to 320GB (40 cores N1 compute instances). It took 4 days of processing (including restarting crashed scripts and restarting pre-emptible machine after 24h window).

**Important tips**:
- install nltk punkt tokenizer, because this component is mandatory and processing pipeline at the end will crash without it:
```
import nltk
nltk.download('punkt')
```
- I have faced strange bug with pickle. During dumping `title2parsed_obj_full_data.p` (pickle.dump was last operation in `blink/candidate_retrieval/enrich_data.py`) process freezed with 100% cpu usage and no new data were appended to `title2parsed_obj_full_data.p`. After few hours of waiting I hit CTRL+C, script exited without any error. Dumped file was not corrupted and next scripts could read it. Here is exact size of file:
```
38322859687 title2parsed_obj_full_data.p
```

Files size in `data/KB_data`:
```
$ du -hs .                                                                       
226G    .
$ du -hs *                                                                       
14G     en_extractor
17G     en_extractor_links
14G     en-wiki
3.3G    en-wiki-filtered
3.9G    en-wiki-filtered-wikidata
14G     en-wiki-full-text
17G     en-wikilinks
37G     en-wikilinks-processed
16G     enwiki-pages-articles.xml.bz2
1.5G    index_enwiki.db
669M    linktitle2wikidataid.p
4.0G    title2enriched_parsed_obj.p
8.9G    title2enriched_parsed_obj_plus.p
36G     title2parsed_obj_full_data.p
42G     wikidata-all.json.bz2
807M    wikidataid_title2parsed_obj.p
318M    wikipediaid2wikidataid.p
```

**Comment by fabiopetroni** [comment]
Thanks @knuser for pointing this out. We are currently in the process of opensourcing a better candidate_retrieval component, that uses much less resources.

**Comment by ledw** [comment]
@knuser We released a newer version which should take less resources. Please try it out if you're interested.
